allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -147,41 +166,72 @@ def forward(self,  # type: ignore
+        num_question_tokens = embedded_question.size(1)
+
+        # (batch_size, num_entities, embedding_dim)
+        encoded_table = self._entity_encoder(embedded_table, table_mask)
+        # (batch_size, num_entities, num_entities)
+        neighbor_indices = self._get_neighbor_indexes(world, num_entities, encoded_table)
+        # (batch_size, num_entities, num_entities, embedding_dim)
+        embedded_neighbors = util.batched_index_select(encoded_table, neighbor_indices)
+        neighbor_mask = util.get_text_field_mask({'ignored': neighbor_indices},
+                                                 num_wrapping_dims=1).float()
+        # (batch_size, num_entities, embedding_dim)
+        embedded_neighbors = self._neighbor_encoder(embedded_neighbors, neighbor_mask)
+
+        # (batch_size, num_entities, num_types)
+        type_vector, entity_types = self._get_type_vector(world, num_entities, encoded_table)
+
+        type_linear_combination = self._type_params(type_vector.float())
+        neighbor_linear_combination = self._neighbor_params(embedded_neighbors.float())
+        # (batch_size, num_entities, embedding_dim)
+        entity_embeddings = torch.nn.functional.tanh(type_linear_combination + neighbor_linear_combination)
+
+        # compute entity and question word similarity through a dot product
+        sim_dot_prod = torch.bmm(embedded_table.view(batch_size, -1, self._embedding_dim),
+                                 torch.transpose(embedded_question, 1, 2))
+        sim_dot_prod = sim_dot_prod.view(batch_size, num_entities, num_entity_tokens, num_question_tokens)
+        # Note the 0th index is accessed below since max returns a tuple of a float and long tensor.
+        # (batch_size, num_entities, num_question_tokens)
+        max_sim = torch.max(sim_dot_prod, 2)[0]
         # (batch_size, num_entities, num_question_tokens, num_features)
-        linking_features = table['linking']  # pylint: disable=unused-variable
-        # TODO(mattg): multiply this by the feature weights, use this and the text embedding above
-        # to compute a linking scores.
+        linking_features = table['linking']
+        linking_scores = max_sim + self._linking_params(linking_features.float()).squeeze(3)
 
-        # TODO(mattg): we need to actually compute this with the stuff mentioned above.  This
-        # represents how well each question token matches table entities.  I'm stubbing this out
-        # for now, so that I have a variable to pass to the decoder, because we need this there.
-        # (batch_size, num_entities, num_question_tokens)
-        num_entities = embedded_table_text.size(1)
-        num_question_tokens = embedded_input.size(1)
-        linking_scores: torch.FloatTensor = Variable(torch.ones(batch_size, num_entities, num_question_tokens))
+        # (batch_size, num_entities+1, num_question_tokens)
+        softmax_scores = self._get_prob_entity(world, linking_scores, batch_size, num_entities,
```

Comment:
Looking at this function, without having seen the method definition, I don't have a good idea of what `get_prob_entity` that returns `softmax_scores` is actually doing.  It's good to give methods names that give the reader a good idea of what the function does, even without having looked at the method.

Having now looked at the method, I would probably call this `_get_linking_probabilities`, `_normalize_linking_scores`, or something similar, with the returned variable just called `linking_probabilities`.  `softmax_scores` could be lots of things, as "softmax" is just a mathematical operation, not something that has semantic content.  It's better to give variables names related to their semantics, not from what operations produced them.

Also, the `linking_scores` tensor already has `batch_size`, `num_entities`, and `num_question_tokens` encoded in its size - passing those variables to this method is redundant, because you can just read them off of `linking_scores` in the method.

Lastly, I think it's more natural to have a normalized tensor like this have its last dimension be the one that's normalized.  So `(batch_size, num_question_tokens, num_entities + 1)` seems a lot more intuitive to me than how you have it.  Your need to do a transpose down below is more evidence of this to me.